### PR TITLE
Fixed type errors in `mypy` GitHub Action

### LINKED ISSUE #6962: Type errors in `mypy` GitHub Action
### What happened?

Running the GitHub Action for [`mypy`](https://github.com/pydata/xarray/blob/main/.github/workflows/ci-additional.yaml#L121) locally returned errors:

```bash
$ python -m mypy --install-types --non-interactive 
xarray/core/merge.py:43: error: "tuple" is not subscriptable  [misc]
xarray/core/merge.py:44: error: "tuple" is not subscriptable  [misc]
xarray/core/merge.py:45: error: "tuple" is not subscriptable  [misc]
xarray/backends/api.py:62: error: "dict" is not subscriptable  [misc]
Found 4 errors in 2 files (checked 144 source files)
```

### What did you expect to happen?

No type errors from `mypy`.

### Minimal Complete Verifiable Example

_No response_

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: 3d3b236df3f8f5982760e68e064eb0db2aacc4a2
python: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10) 
[GCC 10.3.0]
python-bits: 64
OS: Linux
OS-release: 5.10.60.1-microsoft-standard-WSL2
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.8.1

xarray: 2022.6.1.dev39+g3d3b236d
pandas: 1.4.3
numpy: 1.22.4
scipy: 1.9.0
netCDF4: 1.6.0
pydap: None
h5netcdf: 1.0.2
h5py: 3.7.0
Nio: None
zarr: 2.12.0
cftime: 1.6.1
nc_time_axis: 1.4.1
PseudoNetCDF: 3.2.2
rasterio: 1.3.2
cfgrib: 0.9.10.1
iris: 3.2.1
bottleneck: 1.3.5
dask: 2022.8.1
distributed: 2022.8.1
matplotlib: 3.5.3
cartopy: 0.20.3
seaborn: 0.11.2
numbagg: 0.2.1
fsspec: 2022.7.1
cupy: None
pint: 0.19.2
sparse: 0.13.0
flox: 0.5.9
numpy_groupies: 0.9.19
setuptools: 65.3.0
pip: 22.2.2
conda: None
pytest: 7.1.2
IPython: 8.4.0
sphinx: 5.1.1

</details>

### PR DESCRIPTION
**Pull request by lukeconibear**
### Fixed type errors in `mypy` GitHub Action

- [x] Closes #6962

Running the GitHub Action for [mypy](https://github.com/pydata/xarray/blob/main/.github/workflows/ci-additional.yaml#L121) locally returns successful:

```bash
$ python -m mypy --install-types --non-interactive 
Success: no issues found in 144 source files
```

### CODE REVIEW

**Comment by headtr1ck** [comment]
That is weird, with `from __future__ import annotations` this should work.

Actually the pyupgrade pre-commit hook should even replace the typing versions by built-in versions.

**Comment by headtr1ck** [comment]
Could you add an additional mypy workflow to test this behavior?
See https://github.com/pydata/xarray/issues/6962#issuecomment-1277287183

**Comment by headtr1ck** [comment]
When I run locally `mypy --python-version 3.8` i get more errors. Maybe merge current main and let's see what new issues emerge.

**Comment by headtr1ck** [comment]
@dcherian on a different issue: can we prevent that the auto labeler removes manually added labels?

**Comment by dcherian** [comment]
>  can we prevent that the auto labeler removes manually added labels?

no idea. cc @TomNicholas

**Comment by headtr1ck** [comment]
Seems that pyupgrade does not update `tuple` -> `Tuple` inside of `Union`, which is exactly what caused these problems.

Looks good to me now 👍

**Comment by headtr1ck** [comment]
Seems like dasks typing is also not compatible with python 3.8, haha

**Comment by headtr1ck** [comment]
I cannot reproduce the mypy error locally.
Does this break for someone else?

**Comment by max-sixty** [comment]
I get some errors:

```
❯ git describe
v0.17.0-798-g3599f873

❯ mypy --python-version 3.8
xarray/core/types.py:72: error: "tuple" is not subscriptable  [misc]
xarray/core/types.py:74: error: "tuple" is not subscriptable  [misc]
xarray/core/types.py:76: error: "tuple" is not subscriptable  [misc]
xarray/core/types.py:78: error: "list" is not subscriptable  [misc]
xarray/core/_reductions.py:18: error: Cannot find implementation or library stub for module named "flox"  [import]
xarray/core/_reductions.py:18: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
xarray/core/merge.py:43: error: "tuple" is not subscriptable  [misc]
xarray/core/merge.py:44: error: "tuple" is not subscriptable  [misc]
xarray/core/merge.py:45: error: "tuple" is not subscriptable  [misc]
xarray/core/groupby.py:661: error: Cannot find implementation or library stub for module named "flox.xarray"  [import]
xarray/backends/api.py:65: error: "dict" is not subscriptable  [misc]
Found 10 errors in 5 files (checked 139 source files)
```

(not sure why `git describe` is showing that tag but the commit hash should repro)

**Comment by headtr1ck** [comment]
Hmmm, these errors should be fixed in the latest version of this PR.

The problem CI is running into is that mypy runs into an unrecoverable error parsing dask, this I cannot reproduce locally.

**Comment by max-sixty** [comment]
> Hmmm, these errors should be fixed in the latest version of this PR.

Sorry, my test was on main. I can rerun on this PR.

**Comment by max-sixty** [comment]
I can't repro the failure on the current PR `v0.17.0-804-g7018b950`. I'm using Python 3.9 with the latest dask.


```
/home/runner/micromamba/envs/xarray-tests/lib/python3.10/site-packages/dask/utils.py:366: error: Parenthesized context managers are only supported in Python 3.9 and greater  [syntax]
Found 1 error in 1 file (errors prevented further checking)
```

I'm not sure why it's raising an error for an imported module where `--follow-imports=silent`. [Isn't that supposed to suppress errors in imports](https://mypy.readthedocs.io/en/stable/running_mypy.html#follow-imports)?

And weirdly, I get errors running mypy in dask latest main, but not this error!

```
❯ mypy --python-version 3.8 .
dask/system.py:8: error: Unused "type: ignore" comment
dask/widgets/__init__.py:20: error: All conditional function variants must have identical signatures  [misc]
dask/widgets/__init__.py:23: error: All conditional function variants must have identical signatures  [misc]
dask/layers.py:1322: error: Unused "type: ignore" comment
dask/layers.py:1323: error: Unused "type: ignore" comment
dask/array/core.py:5814: error: Argument 1 to "ndindex" has incompatible type "Tuple[int, ...]"; expected "SupportsIndex"  [arg-type]
dask/array/fft.py:196: error: Unused "type: ignore" comment
dask/dataframe/core.py:438: error: Unused "type: ignore" comment
dask/dataframe/core.py:4186: error: Unused "type: ignore" comment
dask/dataframe/core.py:4197: error: Unused "type: ignore" comment
dask/dataframe/core.py:4209: error: Unused "type: ignore" comment
dask/dataframe/core.py:4408: error: Unused "type: ignore" comment
dask/dataframe/core.py:4420: error: Unused "type: ignore" comment
dask/dataframe/core.py:4425: error: Unused "type: ignore" comment
dask/dataframe/groupby.py:1204: error: Unused "type: ignore" comment
dask/dataframe/io/csv.py:9: error: Unused "type: ignore" comment
dask/bytes/tests/test_s3.py:380: error: Unused "type: ignore" comment
dask/bytes/tests/test_local.py:191: error: Unused "type: ignore" comment
dask/bag/tests/test_text.py:33: error: Unused "type: ignore" comment
dask/dataframe/tests/test_shuffle.py:1408: error: Unused "type: ignore" comment
dask/dataframe/tests/test_shuffle.py:1409: error: Unused "type: ignore" comment
dask/dataframe/tests/test_shuffle.py:1412: error: Unused "type: ignore" comment
dask/diagnostics/tests/test_profiler.py:21: error: Unused "type: ignore" comment
dask/tests/test_graph_manipulation.py:106: error: "tuple" is not subscriptable, use "typing.Tuple" instead  [misc]
Found 24 errors in 14 files (checked 243 source files)
```

**Comment by headtr1ck** [comment]
It seems that mypy encountered an unrecoverable runtime error.
We should probably report that to my, but I cannot reproduce it locally, yet alone create a MVCE.

**Comment by Illviljan** [comment]
Dask is on the ignore list:
https://github.com/pydata/xarray/blob/d6671dd414370d006254ba3156cb96256ce0e9c7/pyproject.toml#L31-L43

This seems to ignore the list and follow-imports doesn't seem to work either:
```
      - name: Run mypy with python3.8
        # silent all imports, since external repos might not support this
        run: |
          python -m mypy --install-types --non-interactive --python-version 3.8 --follow-imports=silent
```

A good ol' copy/paste job works as expected though. :) I think we can discuss more elegant solutions in a follow up PR.

**Comment by headtr1ck** [comment]
> A good ol' copy/paste job works as expected though. :) I think we can discuss more elegant solutions in a follow up PR.

So now mypy is not crashing anymore? Thats weird, we should open an issue on mypy about this...

File: .github/workflows/ci-additional.yaml [hunk 1]
```diff
@@ -134,6 +134,64 @@ jobs:
...
+          python xarray/util/print_versions.py
+      - name: Install mypy
+        run: |
+          python -m pip install 'mypy<0.990'
+
+      - name: Run mypy
+        run: |
+          python -m mypy --install-types --non-interactive --cobertura-xml-report mypy_report
+
+      - name: Upload mypy coverage to Codecov
```
Comment (headtr1ck) [inline .github/workflows/ci-additional.yaml:184]: Do we need to upload this to codecov? The coverage should be the same as for the newest python.

File: .github/workflows/ci-additional.yaml [hunk 2]
```diff
@@ -134,6 +134,64 @@ jobs:
...
+        run: |
+          python -m pip install --no-deps -e .
+      - name: Version info
+        run: |
+          conda info -a
+          conda list
+          python xarray/util/print_versions.py
+      - name: Install mypy
+        run: |
+          python -m pip install 'mypy<0.990'
```
Comment (headtr1ck) [inline .github/workflows/ci-additional.yaml:178]: Is mypy=0.990 still crashing?

**Review by headtr1ck — APPROVED** [review]

File: .github/workflows/ci-additional.yaml (hunk 1 above)
Comment (Illviljan) [reply .github/workflows/ci-additional.yaml:184]: Not really, I just kept it simple with the copy/pasting as I suspect we can do this workflow in a smarter way; similar way as the pytest CI or the `mypy --python-version 3.8`-way.

I think we can just focus on getting the xarray bugs fixed in this PR and not get stuck with the CI semantics.

**Comment by Illviljan** [comment]
> > A good ol' copy/paste job works as expected though. :) I think we can discuss more elegant solutions in a follow up PR.
> 
> So now mypy is not crashing anymore? Thats weird, we should open an issue on mypy about this...

This one throws errors still: `python -m mypy --install-types --non-interactive --python-version 3.8 --follow-imports=silent`
Did it ever crash if we used the normal mypy CI but with changed python version?

**Comment by headtr1ck** [comment]
> This one throws errors still: `python -m mypy --install-types --non-interactive --python-version 3.8 --follow-imports=silent`
> Did it ever crash if we used the normal mypy CI but with changed python version?

I don't think we ever used `--python-version` outside of this PR.

We should report this mypy bug...

**Review by headtr1ck — APPROVED** [review]

**Comment by Illviljan** [comment]
Thanks, @lukeconibear !

**Pull request merged**